pkg/istio-agent/xds_proxy.go

```diff
@@ -124,69 +129,116 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
 	if err = proxy.initCertificateWatches(ia, proxy.stopChan); err != nil {
 		return nil, err
 	}
+
+	go proxy.healthChecker.PerformApplicationHealthCheck(func(healthEvent *health.ProbeEvent) {
+		var req *discovery.DiscoveryRequest
+		if healthEvent.Healthy {
+			req = &discovery.DiscoveryRequest{TypeUrl: health.HealthInfoTypeURL}
+		} else {
+			req = &discovery.DiscoveryRequest{
+				TypeUrl: health.HealthInfoTypeURL,
+				ErrorDetail: &google_rpc.Status{
+					Code:    500,
+					Message: healthEvent.UnhealthyMessage,
+				},
+			}
+		}
+		proxy.SendRequest(req)
+	}, proxy.stopChan)
 	return proxy, nil
 }
 
+// SendRequest sends a request to the currently connected proxy
+func (p *XdsProxy) SendRequest(req *discovery.DiscoveryRequest) {
+	p.connectedMutex.RLock()
+	defer p.connectedMutex.RUnlock()
+	// TODO especially for health check purposes, we need a way to ensure the send succeeded. Otherwise,
+	// requests send to a disconnecting proxy will be permanently dropped.
+	if p.connected != nil {
+		p.connected.requestsChan <- req
+	}
+}
+
+func (p *XdsProxy) RegisterStream(c *ProxyConnection) {
+	p.connectedMutex.Lock()
+	defer p.connectedMutex.Unlock()
+	if p.connected != nil {
+		close(p.connected.stopChan)
+	}
+	p.connected = c
+}
+
+type ProxyConnection struct {
+	upstreamError   chan error
+	downstreamError chan error
+	requestsChan    chan *discovery.DiscoveryRequest
+	responsesChan   chan *discovery.DiscoveryResponse
+	stopChan        chan struct{}
+	downstream      discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+}
+
 // Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
 // This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
 // as the new connection may not go to the same istiod. Vice versa case also applies.
 func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
-	proxyLog.Infof("connecting to %s", p.istiodAddress)
+	proxyLog.Infof("Envoy ADS stream established")
+
+	con := &ProxyConnection{
+		upstreamError:   make(chan error),
+		downstreamError: make(chan error),
+		requestsChan:    make(chan *discovery.DiscoveryRequest, 10),
+		responsesChan:   make(chan *discovery.DiscoveryResponse, 10),
+		stopChan:        make(chan struct{}),
+		downstream:      downstream,
+	}
 
-	upstreamError := make(chan error)
-	downstreamError := make(chan error)
-	requestsChan := make(chan *discovery.DiscoveryRequest, 10)
-	responsesChan := make(chan *discovery.DiscoveryResponse, 10)
-	healthEventsChan := make(chan *health.ProbeEvent, 5)
-	// A separate channel for nds requests to not contend with the ones from envoys
-	ndsRequestChan := make(chan *discovery.DiscoveryRequest, 5)
+	p.RegisterStream(con)
```

Comment:
I have the similar doubt with @adiprerepa 

Why donot put the proxyConnection initialization in `initXdsProxy`, and i can not see why `downstream` is necessary. 

And furthermore we can eliminate the lock

Reply:
initXdsProxy initializes a singleton server. This server will have multiple streams (calls to StreamAggregatedResources) established, of which `ProxyConnection` is local to.